packages/flutter/lib/src/widgets/_window_win32.dart

```diff
@@ -0,0 +1,760 @@
+    if (_messageHandlers.contains(handler)) {
+      return;
+    }
+
+    _messageHandlers.add(handler);
+  }
+
+  /// Unregister a [WindowsMessageHandler].
+  ///
+  /// If the handler has not been registered, this method has no effect.
+  ///
+  /// {@macro flutter.widgets.windowing.experimental}
+  ///
+  /// See also:
+  ///
+  ///  * [WindowsMessageHandler], the interface for message handlers.
+  ///  * [WindowingOwnerWin32.addMessageHandler], to register message handlers.
+  @internal
+  void removeMessageHandler(WindowsMessageHandler handler) {
+    _messageHandlers.remove(handler);
+  }
+
+  void _onMessage(ffi.Pointer<_WindowsMessage> message) {
+    final List<WindowsMessageHandler> handlers = List<WindowsMessageHandler>.from(_messageHandlers);
+    final FlutterView flutterView = PlatformDispatcher.instance.views.firstWhere(
+      (FlutterView view) => view.viewId == message.ref.viewId,
+    );
+    for (final WindowsMessageHandler handler in handlers) {
+      final int? result = handler.handleWindowsMessage(
+        flutterView,
+        message.ref.windowHandle,
+        message.ref.message,
+        message.ref.wParam,
+        message.ref.lParam,
+      );
+      if (result != null) {
+        message.ref.handled = true;
+        message.ref.lResult = result;
+        return;
+      }
+    }
+  }
+
+  @internal
+  @override
+  bool hasTopLevelWindows() {
+    return _Win32PlatformInterface.hasTopLevelWindows(PlatformDispatcher.instance.engineId!);
+  }
+}
+
+/// Implementation of [RegularWindowController] for the Windows platform.
+///
+/// {@macro flutter.widgets.windowing.experimental}
+///
+/// See also:
+///
+///  * [RegularWindowController], the base class for regular windows.
+class RegularWindowControllerWin32 extends RegularWindowController
+    implements WindowsMessageHandler {
+  /// Creates a new regular window controller for Win32.
+  ///
+  /// When this constructor completes the native window has been created and
+  /// has a view associated with it.
+  ///
+  /// {@macro flutter.widgets.windowing.experimental}
+  ///
+  /// See also:
+  ///
+  ///  * [RegularWindowController], the base class for regular windows.
+  @internal
+  RegularWindowControllerWin32({
+    required WindowingOwnerWin32 owner,
+    required RegularWindowControllerDelegate delegate,
+    Size? preferredSize,
+    BoxConstraints? preferredConstraints,
+    String? title,
+  }) : _owner = owner,
+       _delegate = delegate,
+       super.empty() {
+    owner.addMessageHandler(this);
+    final ffi.Pointer<_WindowCreationRequest> request = owner.allocator<_WindowCreationRequest>()
+      ..ref.preferredSize.from(preferredSize)
+      ..ref.preferredConstraints.from(preferredConstraints)
+      ..ref.title = (title ?? 'Regular window').toNativeUtf16(allocator: _owner.allocator);
+    final int viewId = _Win32PlatformInterface.createWindow(
+      PlatformDispatcher.instance.engineId!,
+      request,
+    );
+    owner.allocator.free(request);
```

Comment:
`RegularWindowControllerWin32` will become complicated over time as we will add lots of features to it. Should we move all the Flutter Windows FFI down to its own separate layer to simplify `RegularWindowControllerWin32`?

Currently you have `_Win32PlatformInterface.createWindow` which is just a raw FFI API. I'd consider:

1. Rename `createWindow` to `_createWindow`. This is the raw FFI API
2. Add a new `createWindow`. This accepts an `allocator`, `engineId`, `preferredSize`, `preferredConstraints`, and a `title`. It does all allocating + freeing + mapping necessary to call the raw FFI API.